### PR TITLE
get own_dir by using dladdr() instead of argv[0]

### PR DESCRIPTION
**Pull request by soflare**
This PR should fix bug #43

### CODE REVIEW

**Comment by rofl0r** [comment]
should ? or does ?

btw on linux the thing is called "Dl_info", without "struct"

a configure test should be written that checks which one works, then pass it as -DDLINFO_TYPE="struct dlinfo" to the build.

also your patch would be a lot less invasive if you pass dli.dli_fname to the function set_own_dir using the existing parameter.

**Comment by soflare** [comment]
Updated.

I have not tested the #43 since I don't have that environment (seems an unknown version of wine or crossover). But I have 99% confidence about that.

Use `Dl_info` instead of `struct dl_info`, follow the de facto standard prototype of dladdr().
Though dladdr is not an POSIX function, it is implemented in almost all modern *nix system except AIX. If proxychains has AIX support, or has a plan, it makes sense to add a check in configure.

File: src/main.c [hunk 1]
```diff
@@ -49,7 +54,7 @@ static void set_own_dir(const char *argv0) {
 	size_t l = strlen(argv0);
 	while(l && argv0[l - 1] != '/')
 		l--;
-	if(l == 0)
+	if(l == 0 || l >= sizeof(own_dir))
```
Comment (rofl0r) [inline src/main.c]: could you put this into a sep. commit pls ? this seems unrelated to the change at hand

**Comment by rofl0r** [comment]
> Use Dl_info instead of struct dl_info

does apple support that? did you test on mac ?

**Comment by soflare** [comment]
> > Use Dl_info instead of struct dl_info
> 
> does apple support that? did you test on mac ?

Yes, I have tested on mac. FYI, The definition of Dl_info and dladdr() in MacSDK is:
```
#if !defined(_POSIX_C_SOURCE) || defined(_DARWIN_C_SOURCE)
#include <stdbool.h>
#include <Availability.h>
/*
 * Structure filled in by dladdr().
 */
typedef struct dl_info {
        const char      *dli_fname;     /* Pathname of shared object */
        void            *dli_fbase;     /* Base address of shared object */
        const char      *dli_sname;     /* Name of nearest symbol */
        void            *dli_saddr;     /* Address of nearest symbol */
} Dl_info;

extern int dladdr(const void *, Dl_info *);
#endif /* not POSIX */
```

**Comment by rofl0r** [comment]
looks good so far. i just tested this on ubuntu, but it fails:
```
cc src/main.o src/common.o  -o proxychains4
src/main.o: In function `main':
/tmp/proxychains-ng/src/main.c:120: undefined reference to `dladdr'
collect2: error: ld returned 1 exit status
Makefile:89: recipe for target 'proxychains4' failed
make: *** [proxychains4] Error 1
```
we gotta teach the build sys that it needs to add -ldl when linking proxychains4

**Comment by soflare** [comment]
@rofl0r  I have tested on Linux. Unfortunately, `dladdr()` doesn't work. It can only get full path of loaded shared object, but not the main exec.  I changed the patch to use `readlink("/proc/self/exe")`. So, it's no need to add `-ldl` to ldflag.  [On Solaris, the path should be `/proc/self/path/a.out`](https://stackoverflow.com/a/47472873/2186063).

**Comment by rofl0r** [comment]
> I have tested on Linux. Unfortunately, dladdr() doesn't work.

linux != linux

```c
#define _GNU_SOURCE
#include <dlfcn.h>
#include <stdio.h>

char buf[10];

int main(int a, char**b) {
        Dl_info dli;
        dladdr(buf, &dli);
        printf("%s\n", dli.dli_fname);
        printf("%s\n", b[0]);
}
```

this works as expected on linux systems based on musl libc such as sabotage or alpine.
if put into $PATH, it prints full path. on GLIBC-based systems, it returns the same as argv[0].

> I changed the patch to use readlink("/proc/self/exe")

interesting idea.

**Comment by rofl0r** [comment]
i just tested above program on FreeBSD, it works (gives full path).
so i'd say we just use dladdr everywhere, since glibc is the only one not doing it right, in which case nothing changes for the user (as it behaves exactly as the existing code with argv[0])

**Comment by soflare** [comment]
OK, I reverted to the original simple patch. Also, fixed the link error by adding `-ldl` flag.

File: Makefile [hunk 2]
```diff
@@ -86,7 +86,7 @@ $(LDSO_PATHNAME): $(LOBJS)
 		-shared -o $@ $(LOBJS) $(SOCKET_LIBS)
 
 $(ALL_TOOLS): $(OBJS)
-	$(CC) src/main.o src/common.o $(USER_LDFLAGS) -o $(PXCHAINS)
+	$(CC) src/main.o src/common.o $(USER_LDFLAGS) $(LIBDL) -o $(PXCHAINS)
```
Comment (rofl0r) [inline Makefile]: where does LIBDL come from ?

Comment (soflare) [reply Makefile]: The default value of LIBDL is defined at https://github.com/rofl0r/proxychains-ng/blob/master/Makefile#L27. On FreeBSD, the value is cleared at https://github.com/rofl0r/proxychains-ng/blob/master/configure#L210

Comment (rofl0r) [reply Makefile]: oops, missed that :)
